get_global_param.c

```diff
@@ -102,11 +102,41 @@ void ProgramState::initGrid(const std::vector<cell_info_struct>& cells) {
         }
       }
     }
-    global_param.gridNumLatDivisions = ((largestLat - global_param.gridStartLat) / global_param.gridStepLat) + 1;
-    global_param.gridNumLonDivisions = ((largestLon - global_param.gridStartLon) / global_param.gridStepLon) + 1;
+    global_param.gridNumLatDivisions = ((global_param.gridEndLat - global_param.gridStartLat) / global_param.gridStepLat) + 1;
+    global_param.gridNumLonDivisions = ((global_param.gridEndLon - global_param.gridStartLon) / global_param.gridStepLon) + 1;
   }
 }
 
+void ProgramState::initCellMask(const std::vector<cell_info_struct>& cells) {
+
+  char            ErrStr[MAXSTRING];
+	int total_num_cells = global_param.gridNumLatDivisions * global_param.gridNumLonDivisions;
+	valid_cell_mask = new bool [total_num_cells];
```

Comment:
The WriteOutputAllCells::write_data() call to NcVar.putVar assumes that cells are contiguous in a C-language way, i.e. we have to write across all cells (modeled or not modeled) for a given time step, and have no other lat/lon information about them in particular at this stage.  Thus we have to go with the assumption that cells are ordered by increasing lat and then increasing lon.  This requirement has been discussed earlier with regards to multithreading -- ordering of cells in memory is necessary when processing multiple cells in parallel so that their results (which may not arrive in order temporally) can be slotted into the correct position for gathering up and writing out via write_data.  The user is responsible for ordering the soil file cells accordingly.  Alternatively, we could write extra code to accommodate un/incorrectly ordered soil files, and do the sorting during initialization.